### PR TITLE
[HttpFoundation] Add `HeaderRequestMatcher`

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | _NA_
| License       | MIT
| Doc PR        | Todo

This a follow up to:
- https://github.com/symfony/symfony/pull/51324

After @norkunas' [comment](https://github.com/symfony/symfony/pull/51324#issuecomment-1673145964)

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public function matches(Request $request): bool
+    {
+        if (!$this->headers) {
+            return true;
+        }
+
+        foreach ($this->headers as $header) {
+            if (!\in_array($header, $request->headers->keys(), true)) {
```
Comment (norkunas) [inline src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php]: Maybe:

```suggestion
            if (!$request->headers->has($header)) {
```
? 🙂

Comment (alexandre-daubois) [reply src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php]: Much better!

File: src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation\RequestMatcher;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestMatcherInterface;
+
+/**
+ * Checks the HTTP headers of a Request.
```
Comment (derrabus) [inline src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php]: Can we be more precise here? After reading the code, I understood that the class matches request where a given header is simply present. We should point out that we cannot match on the value of headers with this class.

File: src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+class HeaderRequestMatcher implements RequestMatcherInterface
+{
+    /**
+     * @var string[]
+     */
+    private array $headers;
+
+    /**
+     * @param string[]|string $headers A header or a list of headers
+     *                                 Strings can contain a comma-delimited list of headers
```
Comment (derrabus) [inline src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php:31]: Should we really go down that rabbit hole of splitting strings? What would be the use-case?

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php (hunk 2 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php]: That's right, I updated the description :slightly_smiling_face:

File: src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php (hunk 3 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php:31]: Actually, it is to keep consistency with other request matchers, as many other ones accept a comma delimited list

File: src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    public function __construct(array|string $headers)
+    {
+        $this->headers = array_reduce(array_map(strtolower(...), (array) $headers), static fn (array $headers, string $header) => array_merge($headers, preg_split('/\s*,\s*/', $header)), []);
```
Comment (norkunas) [inline src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php]: `strtolower` is not needed I think as request header checking matches case insensitive

Comment (alexandre-daubois) [reply src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php]: You're right! Removed :slightly_smiling_face:

**Comment by alexandre-daubois** [comment]
Status: Needs Review

File: src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php (hunk 3 above)
Comment (GromNaN) [reply src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php:31]: The tests on splitting strings are missing.

**Review by GromNaN — CHANGES_REQUESTED** [review]

**Comment by alexandre-daubois** [comment]
That's right! I added the test here and also in https://github.com/symfony/symfony/pull/51324

File: src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php (hunk 3 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/HttpFoundation/RequestMatcher/HeaderRequestMatcher.php:31]: That's right! I added the test here and also in https://github.com/symfony/symfony/pull/51324

**Comment by alexandre-daubois** [comment]
Rebased and conflicts fixed

**Review by stof — APPROVED** [review]

**Review by 94noni — APPROVED** [review]

**Review by GromNaN — APPROVED** [review]

**Comment by alexandre-daubois** [comment]
Rebased and conflicts fixed :slightly_smiling_face: (as well as [this other related PR](https://github.com/symfony/symfony/pull/51324))

**Review by OskarStark — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alexandre-daubois.

**Pull request merged**